### PR TITLE
Use .jar metadata to generate generic types

### LINKED ISSUE #132: Make asm/jar frontend support java type signatures from .jar file metadata (e.g. for generics)
We should make the asm/jar frontend to `package:jnigen` support java type signatures (see [spec](https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3.4)).

Ideally we would have tests that ensures whether we run `package:jnigen` on `android.jar` or `android-src-stubs.jar` it produces exactly the same code modulo doc comments and parameter names. (one could do that by running jnigen on jar and src, building kernel outlines and comparing the two).

### PR DESCRIPTION
**Pull request by HosseinYousefi**
Closes #132.

### CODE REVIEW

File: jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmClassSignatureVisitor.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  public void visitFormalTypeParameter(String name) {
+    var typeParam = new TypeParam();
+    typeParam.name = name;
+    decl.typeParams.add(typeParam);
+  }
+
+  @Override
+  public SignatureVisitor visitClassBound() {
+    var typeUsage = new TypeUsage();
+    decl.typeParams.get(decl.typeParams.size() - 1).bounds.add(typeUsage);
```
Comment (dcharkes) [inline jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmClassSignatureVisitor.java:31]: Why is it always the last from `decl.typeParams` ?

File: jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmClassSignatureVisitor.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    return new AsmTypeUsageSignatureVisitor(typeUsage);
+  }
+
+  @Override
+  public SignatureVisitor visitInterfaceBound() {
+    var typeUsage = new TypeUsage();
+    decl.typeParams.get(decl.typeParams.size() - 1).bounds.add(typeUsage);
```
Comment (dcharkes) [inline jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmClassSignatureVisitor.java:38]: ditto

File: jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmMethodSignatureVisitor.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  public void visitFormalTypeParameter(String name) {
+    var typeParam = new TypeParam();
+    typeParam.name = name;
+    method.typeParams.add(typeParam);
+  }
+
+  @Override
+  public SignatureVisitor visitClassBound() {
+    var typeUsage = new TypeUsage();
+    method.typeParams.get(method.typeParams.size() - 1).bounds.add(typeUsage);
```
Comment (dcharkes) [inline jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmMethodSignatureVisitor.java:31]: ditto

File: jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmMethodSignatureVisitor.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    return new AsmTypeUsageSignatureVisitor(typeUsage);
+  }
+
+  @Override
+  public SignatureVisitor visitInterfaceBound() {
+    var typeUsage = new TypeUsage();
+    method.typeParams.get(method.typeParams.size() - 1).bounds.add(typeUsage);
```
Comment (dcharkes) [inline jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmMethodSignatureVisitor.java:38]: ditto

File: jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmTypeUsageSignatureVisitor.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        break;
+      case 'S':
+        name = "short";
+        break;
+      case 'V':
+        name = "void";
+        break;
+    }
+    typeUsage.shorthand = name;
+    typeUsage.type = new TypeUsage.PrimitiveType(name);
```
Comment (dcharkes) [inline jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmTypeUsageSignatureVisitor.java:56]: Nit: should we not re-create `PrimitiveType`s for the same `name` every time?

(Feel free to ignore if it needlessly complicates the code.)

File: jnigen/example/in_app_java/jnigen.yaml [hunk 6]
```diff
@@ -14,3 +14,5 @@ source_path:
 classes:
   - 'com.example.in_app_java.AndroidUtils' # from source_path
   - 'android.os.Build' # from gradle's compile classpath
+  - 'java.util.HashMap' # from gradle's compile classpath
```
Comment (dcharkes) [inline jnigen/example/in_app_java/jnigen.yaml:17]: Should we add a separate test instead of adding it on to an example?

If we release the standard library bindings as a separate package, we would likely want to remove HashMap from here again for example purposes. We shouldn't lose test coverage when doing that.

Maybe we should cover tests in `test/` and only add things to `example/` when we really want to illustrate something specific. And then it should probably be accompanied by a README.md detailing what the example shows.

**Review by dcharkes — APPROVED** [review]
lgtm

Thanks @HosseinYousefi !

File: jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmTypeUsageSignatureVisitor.java (hunk 5 above)
Comment (HosseinYousefi) [reply jnigen/java/src/main/java/com/github/dart_lang/jnigen/apisummarizer/disasm/AsmTypeUsageSignatureVisitor.java:56]: I plan to completely change the structure to be more tree like https://github.com/dart-lang/jnigen/issues/157. These will all change then. Parsing that structure would require writing our own parser but it's worth it and we won't have to depend of JsonSerializable.

File: jnigen/example/in_app_java/jnigen.yaml (hunk 6 above)
Comment (HosseinYousefi) [reply jnigen/example/in_app_java/jnigen.yaml:17]: Good idea to separate the example. Will do that.

**Comment by HosseinYousefi** [comment]
The issue with the test is interesting. We have a class that is defined like:

```java
public class JsonFactoryBuilder extends TSFBuilder<JsonFactory, JsonFactoryBuilder>
```

So it extends a class which has itself as a type parameter. Although the descriptor for this class is `com/fasterxml/jackson/core/JsonFactoryBuilder`, the signature becomes `Lcom/fasterxml/jackson/core/TSFBuilder<Lcom/fasterxml/jackson/core/JsonFactory;Lcom/fasterxml/jackson/core/JsonFactoryBuilder;>;`. This edge cases causes the asm backend test to fail.

File: jnigen/test/jackson_core_test/generate.dart [hunk 7]
```diff
@@ -56,13 +56,16 @@ Config getConfig(
...
     exclude: BindingExclusions(
-      fields: excludeAll<Field>([
-        ['com.fasterxml.jackson.core.JsonFactory', 'DEFAULT_QUOTE_CHAR'],
-        ['com.fasterxml.jackson.core.Base64Variant', 'PADDING_CHAR_NONE'],
-        ['com.fasterxml.jackson.core.base.ParserMinimalBase', 'CHAR_NULL'],
-        ['com.fasterxml.jackson.core.io.UTF32Reader', 'NC'],
-      ]),
-    ),
+        fields: excludeAll<Field>([
+          ['com.fasterxml.jackson.core.JsonFactory', 'DEFAULT_QUOTE_CHAR'],
```
Comment (dcharkes) [inline jnigen/test/jackson_core_test/generate.dart:60]: Do we need a TODO for these?

Comment (HosseinYousefi) [reply jnigen/test/jackson_core_test/generate.dart:60]: These were existing before. Maybe?!
